docs/vocabulary.md

```diff
@@ -388,6 +388,18 @@ pipeline in corresponding documentation.
 | **Tekton**         | N/A       | Step       | Task     | Pipeline | Trigger | Resource (?)           |
 | **Zuul**           | N/A       | N/A        | Job      | Pipeline | Trigger | Node (?)               |
 
+### Build Stages
+
+The table below is an attempt to create a mapping of common names for build stages (the unit of work one degree smaller than a pipeline). While CI/CD Tools and Technologies generally give developers broad leeway in naming and implementing their build stages, agreeing upon some common terms will help develop more pluggable build pipelines.
+
+| Stage      | Semantics     | Aliases       | Inputs   | Outputs and Results  |
+|------------|---------------|---------------|----------|-----------------------|
+| **Build** | Set up the build workspace and provision build resources. Download, retrieve, assemble, and/or compile software into an executable and testable format. Download, retrieve, assemble, and/or compile documentation into a consumable format. | N/A | Source Code, Other Software Artifacts, Documentation | Executable Software, Compiled Documentation |
```

Comment:
To the list of inputs I would add "Baseline" or "Composition", i.e. the actual _list(s)_ of upstream source code, documents and/or artifacts, all with their specific versions, that are brought in to the actual build. In some scenarios such a baseline/composition is part of the source code repo, in some it is calculated as a pre-step in the build process based on makefiles and such, and in some scenarios it is created by some external service from where the build process would retrieve it as input.

If the baseline/composition is generated within the build step it could also be considered an output, i.e. compiled piece of information.

Output from the build step, and all other pipeline steps, should probably also include "Logs" and possibly also other compiled, non-executable pieces of information.

Reply:
Does the "Baseline" or "Composition" include the Dependencies, then? Or do you think I should list that separately?

Maybe we can factor out the items that are included across all stages, like Logs and Secrets. That could be another row in the table, or notes above or below it.

Reply:
I re-formatted most of this into bulleted lists because, according to some style guides, tables should always be compact and simple data.  The semantics and results of the various stages were making the table pretty dense and wordy.

Reply:
Baselines/Compositions are instances of the Dependencies, to me. Dependencies would be a list of upstream items/artifacts to integrate or use in the current stage, but most often without explicit versions. They could probably specify ranges of versions though. A Baseline/Composition should always have specific versions for all its items, so that re-running the same stage with that Baseline/Composition as input would render the same output.

Reply:
OK, so using Node.js as an example, you would consider the package.json file that lists the dependencies (and may use "latest" or a major version) an example of "Dependencies".  I'll use the term "Declared Dependencies" in the next revision to make that more obvious.

The Baseline/Composition would be the package.lock file with the exact versions and hashes for each package.  Normally the Test, Release, and Deploy stages would use the package.lock file to guarantee consistency.  You can also commit package.lock to SCM to get the Build stage to use the same package versions as well.  Sound right?

Reply:
I think Declared Dependencies (what the user asks for) is a good term.  The Baseline/Composition *could* be (for rebuild and consistency ala package-lock.json) but I think it needs to be captured as a computed output (npm update rebinds the dependencies and as your use case can be committed and used next time if no updates are wanted). The Baseline/Composition is computed (example is package-lock.json) as it contains the exact versions of what I sometimes call "As-Built" Dependencies (what the build actually found and used when interpreting the Declared Dependencies (ie latest) . This Baseline/Composition is also input to the SBOM and the content generated by a build as the most accurate list of components the build actually used.  So, I think we need to consider "lock" type output as part a generated of part of a build, specifically to capture any computed dependencies aka the "as-built" parts.  package-lock.json is a good example, and maven ranges (need to determine which one actually matches). There have also been Dockerfile.lock type solutions proposed which creates a new Dockerfile with tags replaced with precise versions (ie FROM image:latest -> FROM image:exact-sha-tag) so that a rebuild is exact.  I would say that this concept is also needed in container deployment scenarios (which can reference imprecise tags/labels) and any artifacts that are input, also need generated "lock" versions . The additional outputs that build also *could* produce are versions of tools (compiler versions, package manager versions, task container versions in tekton, OS versions etc) which captures some details on environment which is that point in time state needed for anyone who wants to rebuild or understand what went into a build.

Reply:
I added info about those additional outputs.  The Declared Dependencies/Baseline/Composition comments are refactored into the definitions of "Software Source" and "Generated Software" now.